tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc

```diff
@@ -54,6 +54,74 @@ __device__ __forceinline__ void AccumulateInto(
+// UnsortedSegmentSumCustomKernel does except that input data
+// is partitioned along the outer reduction dimension. This is
+// because consecutive rows (elements in a row share the same
+// outer dimension index) in the flattened 2D input data likely
+// belong to the same segment in sorted segment sum operation.
+// Therefore such partitioning strategy has two advantages over
+// the UnsortedSegmentSumFunctor kernel:
+// 1. Each thread reduces across multiple rows before writing
+// answers to the global memory, we can therefore
+// write reduction results to global memory less often.
+// 2. We may know that the current thread is the only contributor
+// to an output element because of the increasing nature of segment
+// ids. In such cases, we do not need to use atomic operations
+// to write results to global memory.
+// In the flattened view of input data (with only outer and inner
+// dimension), every thread processes a strip of input data of
+// size OUTER_DIM_TILE_SIZE x 1. This strip runs across multiple
+// rows of input data and all reduction elements share one inner
+// dimension index.
+#define OUTER_DIM_TILE_SIZE 8
+#define CEIL_DIV(x, y) (1 + (((x)-1) / (y)))
+template <typename T, typename Index>
+__global__ void SortedSegmentSumCustomKernel(const Index input_outer_dim_size,
+                                             const Index inner_dim_size,
+                                             const Index output_outer_dim_size,
+                                             const Index* segment_ids,
+                                             const T* input, T* output) {
+  const Index input_outer_dim_num_strip =
+      CEIL_DIV(input_outer_dim_size, OUTER_DIM_TILE_SIZE);
+  const Index total_stripe_count = inner_dim_size * input_outer_dim_num_strip;
+
+  CUDA_1D_KERNEL_LOOP(strip_index, total_stripe_count) {
+    const Index segment_offset = strip_index % inner_dim_size;
+    const Index input_outer_dim_index_base =
+        strip_index / inner_dim_size * OUTER_DIM_TILE_SIZE;
+
+    T sum = T(0);
+    Index first_segment_id = segment_ids[input_outer_dim_index_base];
+    Index last_output_segment_id = output_outer_dim_size;
+    const Index actual_stripe_height = MIN(
+        OUTER_DIM_TILE_SIZE, input_outer_dim_size - input_outer_dim_index_base);
+    for (Index j = 0; j < actual_stripe_height; j++) {
+      Index current_output_segment_id =
+          segment_ids[input_outer_dim_index_base + j];
+      // decide whether to write result to global memory
+      if (current_output_segment_id > last_output_segment_id) {
+        const Index output_index =
+            last_output_segment_id * inner_dim_size + segment_offset;
+        // decide whether to write result to global memory using atomic
+        // operations
+        if (last_output_segment_id == first_segment_id) {
+          AccumulateInto<T>(output + output_index, sum);
+        } else {
+          *(output + output_index) = sum;
```

Comment:
I took a closer look at this code. I think we should definitely add more tests to make sure there is no hidden race condition. Also adding a benchmark showing there is no performance regression against unsorted_segment_sum is good. 

Please make sure try a few different tensor shape aspect ratio in the benchmarks.